src/Service/DonationService.php

```diff
@@ -0,0 +1,127 @@
+
+class DonationService
+{
+    const PAYMENT_METHOD_TYPE_SCHEME = 'scheme';
+    const SHOPPER_INTERACTION_CONTAUTH = 'ContAuth';
+
+    /**
+     * @var ClientService
+     */
+    private $clientService;
+
+    /**
+     * @var ConfigurationService
+     */
+    private $configurationService;
+
+    /**
+     * @param ClientService $clientService
+     * @param ConfigurationService $configurationService
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        ClientService $clientService,
+        ConfigurationService $configurationService,
+        LoggerInterface $logger
+    ) {
+        $this->clientService = $clientService;
+        $this->configurationService = $configurationService;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param SalesChannelContext $context
+     * @param $donationToken
+     * @param $currency
+     * @param $value
+     * @param $returnUrl
+     * @param $pspReference
+     * @return array|mixed
+     * @throws \Adyen\AdyenException
+     */
+    public function donate(SalesChannelContext $context, $donationToken, $currency, $value, $returnUrl, $pspReference)
+    {
+        $responseData = [];
+
+        $requestData = $this->buildDonationRequest(
+            $context,
+            $donationToken,
+            $currency,
+            $value,
+            $returnUrl,
+            $pspReference
+        );
+
+        if (!empty($requestData)) {
+            $checkoutService = new CheckoutService(
+                $this->clientService->getClient($context->getSalesChannel()->getId())
+            );
+            $responseData = $checkoutService->donations($requestData);
+        }
+
+        return $responseData;
+    }
+
+    /**
+     * @param SalesChannelContext $context
+     * @param $donationToken
+     * @param $currency
+     * @param $value
+     * @param $returnUrl
+     * @param $pspReference
+     * @return array
+     */
+    public function buildDonationRequest(
+        SalesChannelContext $context,
+        $donationToken,
+        $currency,
+        $value,
+        $returnUrl,
+        $pspReference
+    ) : array {
+        $request['amount']['currency'] = $currency;
+        $request['amount']['value'] = $value;
+        $request['reference'] = Uuid::randomHex();
+        $request['donationToken'] = $donationToken;
+        $request['donationOriginalPspReference'] = $pspReference;
+        $request['donationAccount'] = $this->configurationService->getAdyenGivingCharityMerchantAccount(
+            $context->getSalesChannel()->getId()
+        );
+        $request['merchantAccount'] =  $this->configurationService->getMerchantAccount(
+            $context->getSalesChannel()->getId()
+        );
+        $request['paymentMethod']['type'] = self::PAYMENT_METHOD_TYPE_SCHEME;
+        $request['shopperInteraction'] = self::SHOPPER_INTERACTION_CONTAUTH;
+        $request['returnUrl'] = $returnUrl;
+
+        return $request;
```

Comment:
How about structuring this as something like ...
```suggestion
        return [
            'amount' => [
                'currency' => $currency,
                'value' => $value
            ],
            'reference' => Uuid::randomHex(),
            'donationToken' => $donationToken,
            ....
        ];
```